### PR TITLE
Demo video

### PR DESCRIPTION
**Pull request by KomlaRD**
# Description

- Changed start script in the package.json file. Current start script makes server restart whenever changes are made unlike the previous start script.
- Added an index route to act as the main route.
- Added a video route to stream videos from the backend.
- Added three sample cropped videos in the assets folder.
- Added a metadata file to provide info of videos in the assets folder.
- Added unit testing for endpoints

# Testing

- Testing with Jasmine

Screenshot 2023-05-19 at 7 14 38 PM

- Server working as expected

Screenshot 2023-05-19 at 9 05 16 PM

- Metadata for all videos in the assets folder

Screenshot 2023-05-19 at 9 05 31 PM

- Metadata for single video

Screenshot 2023-05-19 at 9 07 42 PM

- Demo video

https://github.com/nsmq-ai/nsmqai/assets/72665568/69b96fa5-37a0-42e7-9058-0a713dd9c5ac

# Reviewers

@wsedor @AmoabaKelvin

### CODE REVIEW

**Comment by wsedor** [comment]
@KomlaRD can you please add some unit testing for these changes?

File: webapp/backend/src/handlers/api/videos.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import { Request, Response, Router } from "express";
+import fs from "fs";
+import allVideos from "../../videoData";
+import path from "path";
+
+// Initialise router object
+const demoVideos = Router();
```
Comment (wsedor) [inline webapp/backend/src/handlers/api/videos.ts:7]: let's keep the name of this variable as `routes` to be consistent with the naming in `index.ts` file

File: webapp/backend/src/handlers/api/videos.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+// Metadata for all videos
+demoVideos.get("/", (req: Request, res: Response) => {
+  res.json(allVideos);
+});
+
+// Metadata for a single video
+demoVideos.get("/:id/metadata", (req: Request, res: Response) => {
```
Comment (wsedor) [inline webapp/backend/src/handlers/api/videos.ts:15]: do some error checking of the user input here. shouldn't implicitly trust that it is correct or safe

File: webapp/backend/src/handlers/api/videos.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  const id = parseInt(req.params.id, 10);
+  res.json(allVideos[id]);
+});
+
+// Endpoint for a single video to stream
+demoVideos.get("/video/:id", (req: Request, res: Response) => {
```
Comment (wsedor) [inline webapp/backend/src/handlers/api/videos.ts:21]: We already have the parent `/videos` level which is the collection URI. We don't want to unnecessarily make the api endpoint too long. Let's follow the REST api naming convention below and go with `/videos/{id}` for a particular video instead of `/videos/video/{id}`

https://restfulapi.net/resource-naming/

Comment (wsedor) [reply webapp/backend/src/handlers/api/videos.ts:21]: please go over the api conventions as well for future use cases also.

Comment (wsedor) [inline webapp/backend/src/handlers/api/videos.ts:21]: do some error checking of the id passed here. shouldn't implicitly trust that it is correct or safe

File: webapp/backend/src/handlers/index.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+// Router object
+const routes = Router();
+
+// Main route
+routes.get("/", (req: Request, res: Response) => {
+    res.send("Welcome to NSMQ AI Web backend")
+});
+
+// Demo video route
+routes.use("/videos", demoVideos);
```
Comment (wsedor) [inline webapp/backend/src/handlers/index.ts:13]: we will also have live video endpoints so let's use `demo-videos` or more applicable endpoint name to differentiate them?

**Comment by wsedor** [comment]
@KomlaRD can you also add a few seconds of video demonstrating that the functionality works in the example html you used in development. Can use a screen recorder for this purpose

**Comment by KomlaRD** [comment]
> @KomlaRD can you please add some unit testing for these changes?

Sure. I will work on that.

File: webapp/backend/src/handlers/api/videos.ts (hunk 1 above)
Comment (KomlaRD) [reply webapp/backend/src/handlers/api/videos.ts:7]: > let's keep the name of this variable as `routes` to be consistent with the naming in `index.ts` file

I thought the names had to be different... or do you mean something like `demoVideoRoutes` or just `routes`?

File: webapp/backend/src/handlers/api/videos.ts (hunk 2 above)
Comment (KomlaRD) [reply webapp/backend/src/handlers/api/videos.ts:15]: > do some error checking of the id passed here. shouldn't implicitly trust that it is correct or safe

Sure. I will work on that. Thanks for pointing that out.

File: webapp/backend/src/handlers/api/videos.ts (hunk 3 above)
Comment (KomlaRD) [reply webapp/backend/src/handlers/api/videos.ts:21]: > please go over the api conventions as well for future use cases also.

Sure. Thanks for the additional resource.

File: webapp/backend/src/videoData.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+const allVideos = [
```
[the remark this answers is not in the document]
Comment (KomlaRD) [reply webapp/backend/src/videoData.ts:1]: > can you name this all `allDemoVideos` or more applicable name?

Yes please. Thanks.

File: webapp/backend/src/handlers/index.ts (hunk 4 above)
Comment (KomlaRD) [reply webapp/backend/src/handlers/index.ts:13]: > we will also have live video endpoints so let's use `demo-videos` or more applicable endpoint name to differentiate them?

Sure. Thanks for the feedback.

**Comment by KomlaRD** [comment]
> @KomlaRD can you also add a few seconds of video demonstrating that the functionality works in the example html you used in development? Can use a screen recorder for this purpose

Okay. Sure.

File: webapp/backend/src/handlers/api/videos.ts (hunk 1 above)
Comment (wsedor) [reply webapp/backend/src/handlers/api/videos.ts:7]: why do the names have to be different? the return of the Router() object is not specific to the demo routes is it? isn't it just a route object?

Comment (KomlaRD) [reply webapp/backend/src/handlers/api/videos.ts:7]: Sure, it is just a route object. I will make the naming consistent as suggested. My response was a thought I had about routes. Thanks for clarifying.

File: webapp/backend/src/handlers/api/tests/videos.spec.ts [hunk 6]
```diff
@@ -29,8 +29,8 @@ describe("Demo video routes", () => {
 
   it("should get response from demo videos metadata with id", async () => {
     try {
-        const response = await request.get(`/${id}/metadata`);
-        expect(response.status).toBe(200);
+        const response = await request.get(`/${id})/metadata`);
+        expect(response.status).toBe(400);
```
Comment (wsedor) [inline webapp/backend/src/handlers/api/tests/videos.spec.ts:33]: this looks like a negative test? if so please modify the test description on line 30

File: webapp/backend/src/handlers/api/videos.ts [hunk 7]
```diff
@@ -26,6 +37,11 @@ demoVideosRoutes.get("/:id", (req: Request, res: Response) => {
         // Request parameter id
         const id = req.params.id;
 
+        // Check if id is a non-empty string
```
Comment (wsedor) [inline webapp/backend/src/handlers/api/videos.ts:40]: do we not need similar checks for id that were used in metadata section here https://github.com/nsmq-ai/nsmqai/pull/18/commits/0f8ded90f2edd984ed60b5dc84e8869893ba8788#diff-5fd61999e79dcd8ccc12ecb6b6fbfda464fb914306e10ed042e9d6e6b9bcdabcR19? 

You can extract out the checks for id into a function and use it both in the metadata param endpoint and also the single video param endpoint checking as well

File: webapp/backend/src/handlers/api/tests/videos.spec.ts (hunk 6 above)
Comment (wsedor) [reply webapp/backend/src/handlers/api/tests/videos.spec.ts:33]: please add tests for empty, `NaN` and negative id as well since they are checks in the code in `videos.ts`

**Review by wsedor — APPROVED** [review]
